const { User } = require('../models/index.js');

//Come√ßo de uma verifica√ß√£o por cargo bem mais complexa, do que precisa -se para este tipo de app
// const checkPermissions = (req, res, perm, next) => {

//     let login = req.session.user

//     if (!login){
//         return res.status(401).send('N√£o autorizado. Fa√ßa login primeiro.');
//     }

//     login = User.findOne({ where: { ramal: login } });

//     if(!login){
//         return res.status(500).send('N√£o autorizado. Fa√ßa login primeiro.');
//     }

//     console.log(login.role);

// }

//  Puxar o cargo do user.
// Puxar as permiss√µes atreladas ao cargo do user
// Check the true or false
// If true let's go
// Else can return false

const checkPermissions = (requiredRole) => {
    return async (req, res, next) => {
        const login = req.session.user;

        if (!login) {
            return res.status(401).json({
                success: false,
                message: 'N√£o autorizado. Fa√ßa login!'
            });
        }

        try {
            const user = await User.findOne({
                where: { id: login.id }
            });

            if (!user) {
                return res.status(401).json({
                    success: false,
                    message: 'Usu√°rio n√£o encontrado'
                });
            }

            if (user.role === requiredRole) {
                req.user = user; // Adicionei para disponibilizar o usu√°rio no request
                return next();
            } else {
                return res.status(403).json({
                    success: false,
                    message: 'üîí Acesso restrito! Voc√™ n√£o possui as permiss√µes necess√°rias.'
                });
            }

        } catch (error) {
            console.error('Erro na verifica√ß√£o de permiss√µes:', error);
            return res.status(500).json({
                success: false,
                message: 'Erro interno no servidor'
            });
        }
    };
};


module.exports = checkPermissions;